documentation/tut/posting_data.rst

```diff
@@ -269,29 +269,32 @@ Posting flexibility states
 
 There is one more crucial kind of data that FlexMeasures needs to know about: What are the current states of flexible devices?
 For example, a battery has a certain state of charge, which is relevant to describe the flexibility that the battery currently has.
+In our terminology, this is called the "flex model" and you can read more at `ref`:describing_flexibility:.
 
-Owners of such devices can post these states along with triggering the creation of a new schedule, to `[POST] /schedules/trigger <../api/v3_0.html#post--api-v3_0-sensors-(id)-schedules-trigger>`_.
+Owners of such devices can post the flex model along with triggering the creation of a new schedule, to `[POST] /schedules/trigger <../api/v3_0.html#post--api-v3_0-sensors-(id)-schedules-trigger>`_.
 The URL might look like this:
 
 .. code-block:: html
 
     https://company.flexmeasures.io/api/<version>/sensors/10/schedules/trigger
 
-This example triggers a schedule for a power sensor (with ID 10) of a battery asset, asking to take into account the battery's current state of charge.
+The following example triggers a schedule for a power sensor (with ID 10) of a battery asset, asking to take into account the battery's current state of charge.
 From this, FlexMeasures derives the energy flexibility this battery has in the next 48 hours and computes an optimal charging schedule.
-The endpoint allows to limit the flexibility range and also to set target values.
+The endpoint also allows to limit the flexibility range and also to set target values.
 
 .. code-block:: json
 
         {
-            "value": 12.1,
-            "datetime": "2015-06-02T10:00:00+00:00",
-            "unit": "kWh"
+            "start": "2015-06-02T10:00:00+00:00",
+            "flex-model": {
+                "soc_at_start": 12.1,
+                "soc_unit": "kWh"
```

Comment:
Actually no, at this time the storage flex model uses underscores for its parameters. I would like to have the outside world use ashes, but us using a schema not immediately on the input makes this difficult.

Here is why we do that: We allow some fields to be empty, and be filled in from sensor attributes (e.g. `soc_min`, `roundtrip_efficiency`). The Marshmallow schema validation does not have access to the sensor, and so I have code to check and fill in that information *before* the schema validation.

If you have an idea how we can let schema validation know the sensor, I'm all ear :)

Reply:
Dashes are preferred, but that's it. Reading your review, I wonder how you thought my tests could have been passing :)

People can write flex models with parameters named in any way.

Reply:
I don't fully understand yet how these two issues relate. If the schema validation (which one?) doesn't have the sensor object, we cannot use dashes in field names within the `flex-model` and `flex-context` fields: why is that?

It seems to me that the fields `flex-model` and `flex-context` are deserialized (by the validators on `SensorAPI.trigger_schedule`) to `dict` variables `flex_model` and `flex_context`, then stored as scheduler attributes. They are just dictionaries at this point, so their keys can contain dashes. When calling `inspect_flex_config` on the scheduler, missing fields are filled in and the flex model and context are finally each deserialized. At this point I believe it makes sense to deserialize the dict keys to become underscored, so that the dict could be used as kwargs (we don't use them that way btw, in `StorageScheduler.compute_schedule` we just use `get` on the dict, and the keys could still be dashed, but I agree that at that point in the code one would expect underscored keys that look like Python variables). To make this happen, the fields in `StorageFlexModelSchema` could each get a dashed `data_key`.

My opinion: mixing dashes and underscores in the API is a bad practice. We should refrain from it, and discourage it for people writing their own flex models.

Reply:
I looked up if marshmallow will honor the data key in the model after deserialization, but it doesn't. So after that step, developers have to use underscores when accessing the flex_model dict, before they still have to use dashes.
That confusing situation led me to this decision. We can decide that coherent API style is more important than this dev confusion, though.